### ISSUE TITLE
Running websockproxy without TUN/TAP

### ISSUE THREAD

**Issue by benjamincburns** [benjamincburns/websockproxy#23]
### Discussed in https://github.com/benjamincburns/websockproxy/discussions/22

<div type='discussions-op-text'>

<sup>Originally posted by **SSP6904** March  1, 2023</sup>
Hi!

I'm trying to do this in my v86 repo, along with the Docker file in the root.
But it takes a while or if that's just me..

Is there a version that can deploy on GitHub actions?</div>

Unfortunately at the moment there is no way to run `websockproxy` as a standalone virtual switch. At the moment, `websockproxy` always expects to be able to allocate and connect to a TAP device. This means that it simply won't run in unprivileged environments, such as in a docker container that's created without the `--privileged` flag.

It should be easy however to make `websockproxy` fall back to being a dumb switch in cases when it fails to allocate or access a TAP device. This will get around the need to run it in a privileged context. Unfortunately this may create new problems for some use cases, however. Specifically, setups relying on `websockproxy` transitively depend on the operating system and other userspace daemons to provide the following features:

- DHCP
- Hostname allocation and DNS
- NAT gateway (aka accessing the internet)

It is likely possible to bake some of these features into `websockproxy` directly, but I consider those to be out of scope for this ticket.

**Comment by benjamincburns** [comment]
> 🤖 **This comment was written by an AI agent (Claude Code), at @benjamincburns's request.** It summarises a design discussion about expanding this issue's scope. Where it says something was tested or verified, that was done in the agent's session; everything else is a proposal.

## Proposal: widen the scope from "switch without TAP" to "self-contained router without TAP"

As written, this issue asks for a TAP-less **switch**, and puts DHCP, DNS and NAT out of scope. This comment proposes bringing them in scope: websockproxy would run as a self-contained, **software-defined router** that needs no TUN/TAP device, no `--privileged` container, no iptables and no dnsmasq.

That's a well-established approach. QEMU's default "user-mode" networking (libslirp), passt, VirtualBox NAT and gVisor's netstack all work this way. Guest frames addressed to the virtual gateway are handled by a network stack running *inside the relay*. It reaches the outside world through ordinary, unprivileged sockets on the guests' behalf.

### Why it's worth doing

- **No privileged container.** Today the relay must run with `--privileged`, so any bug in it runs with full privileges. This would be the biggest security improvement available to the project.
- **Access rules in code.** Rules like "public addresses only, plus these allowlisted private ranges" become a check in the relay before each outbound connection, instead of iptables rules in a startup script. They're unit-testable and don't depend on how the host is set up.
- **It runs in far more places,** including environments where TUN/TAP isn't available. That's what the Discussion behind this issue was after. (One caveat: GitHub Actions itself isn't a host for long-lived public servers, so this won't make *that* specific use case work.)
- **Better abuse handling.** The relay sees every flow, so it can attribute traffic to a client (IP ↔ MAC ↔ DHCP lease) and limit it per client.

### Trade-offs

- **Only TCP, UDP and ICMP are carried.** Other IP protocols (GRE, SCTP, raw sockets) won't work. That's fine for the intended use (jor1k/v86 guests browsing, running `apt`, `curl`, `git`), but it isn't a full router.
- **No inbound connections** unless explicit port forwarding is added later.
- **More resources per guest.** Every guest connection or flow becomes a host socket. That calls for per-client limits on flows and sockets, idle timeouts and a global file-descriptor budget. The existing per-frame rate limits aren't enough on their own.
- **Security exposure moves into the relay.** It would be parsing and holding state for untrusted IP, TCP and UDP traffic from guests. A mature stack is strongly preferable to writing one (see options below). Even mature stacks have a CVE history (libslirp has had several), so the dependency must be pinned and kept up to date.
- **Per-packet CPU cost in the relay.** Probably not noticeable, since the websocket hop is already the bottleneck (see #4).
- **Still Linux-only.** Some of the ICMP support below relies on Linux-specific socket features. The relay is already Linux-only, so this isn't a regression.
- **TAP mode stays.** It remains the choice for anyone who wants full kernel routing (all protocols, real global IPv6, inbound connections).

## What the relay would need

### Groundwork: a swappable uplink
Today the relay's uplink is always a TAP device. Refactor it into a pluggable "port" that sends and receives Ethernet frames, selected with `WEBSOCKPROXY_UPLINK`:

- `tap`: current behaviour.
- `none`: pure switch; this issue's original scope.
- `router`: the userspace router described below.

Shipping `none` first is a useful milestone on its own.

### IPv4

| Piece | What it does | Effort |
|---|---|---|
| ARP | Answer for the virtual gateway (e.g. `10.5.0.1`) with a virtual MAC | Trivial |
| DHCP | Hand out addresses from a pool, keyed by MAC. The switch already tracks which client owns each MAC, so a lease can be released when its client disconnects | Small |
| DNS | Forward guest queries to the host's resolver over an ordinary UDP socket (in Docker that's whatever `/etc/resolv.conf` says) | Small |
| UDP NAT | One host UDP socket per guest flow, with a flow table and idle timeouts | Small to medium |
| **TCP NAT** | End each guest TCP connection inside the relay and open a normal host socket to the real destination. That needs a real guest-side TCP implementation: handshake, sequencing, retransmission, windows, closing and reset. Flow control ties into the existing downstream rate limit and `MAX_PENDING_SENDS` | **The bulk of the work** |
| Fragment reassembly | For guest UDP datagrams larger than the MTU | Small |
| Access rules | The same rules as today's TAP-mode firewall, checked on the actual destination IP before each connection: block IANA special-purpose ranges, with `WEBSOCKPROXY_ALLOWED_PRIVATE_NETS` for exceptions. Because the check is on the IP the guest actually connects to, DNS tricks (DNS rebinding) can't get around it | Small |

### ICMP
- **Pinging the gateway:** the relay answers these itself.
- **Pinging the internet:** use unprivileged ICMP "ping sockets". The kernel rewrites the echo ID, so the relay keeps a table mapping it back to the guest's ID and sequence number.
  - **Tested:** ping sockets work in a default Docker container, even as user `nobody` with `--cap-drop ALL`, because Docker allows every group to use them (`ping_group_range = 0 2147483647`).
  - **Bare-host installs** often have them disabled (`1 0`), so the operator would need to widen that setting. Otherwise ping from guests to the internet is unavailable, and everything else still works.
- **Errors for the guest's own traffic** (unreachable, time exceeded, fragmentation needed): enable the `IP_RECVERR` socket option on each outbound socket and read ICMP errors from the socket's error queue (`MSG_ERRQUEUE`). That gives the ICMP type and code, the router that sent it, and the path MTU. The relay then builds a matching ICMP error for the guest, quoting the *guest's* original packet header.
  - **Tested:** a UDP packet sent with TTL=1 from an unprivileged container produced a Time Exceeded (type 11) from the first-hop router in the error queue, with the original payload.
  - **Python gap:** Python ≤3.12 doesn't define `socket.IP_RECVERR`, so the code uses its value (`11`) directly.
- **Traceroute:** copy the guest's TTL minus one onto each outbound probe (`IP_TTL`). Answer "time exceeded" from the gateway when a packet arrives with TTL 1, so the relay shows up as hop 1. This covers UDP, ICMP and TCP traceroute.
- **TCP errors:** a refused connection becomes a TCP reset to the guest, and an unreachable host becomes an ICMP "host unreachable".
- **Privacy:** hops with non-public addresses on the relay's own path (the Docker host, the operator's router, VPN hops) would otherwise show up in guests' traceroutes. Rewrite or hide them by default.
- **Abuse limits:** rate-limit ICMP per client (pings, and error replies the relay generates) and cap ping sockets per client.
- **Everything else** (redirects, timestamps, router discovery, unsolicited inbound ICMP) is dropped.

### IPv6
- **The switch already handles IPv6.** Multicast MACs (`33:33:…`) are flooded to all clients, and duplicate-address detection works now that broadcasts aren't reflected back to their sender.
- **Neighbor Discovery** (the IPv6 equivalent of ARP) answers for the gateway's addresses.
- **Router advertisements** give guests a prefix to configure their own addresses from (SLAAC), and include the DNS server (the RDNSS option, RFC 8106). Stateless DHCPv6 is optional, for guests that can't read DNS from advertisements.
- **Addressing:** a random ULA `/64` generated at startup (RFC 4193). Traffic leaves through the host's own IPv6 address, which behaves like NAT66, and needs no host configuration.
  - **Consequence:** under the standard address-selection rules (RFC 6724), guests will prefer IPv4 for destinations that have both, and use IPv6 only for IPv6-only destinations. That's harmless, but IPv6 won't be the default path.
  - **Real global addresses for guests** need a prefix routed to the host and can't work through socket proxying, so they'd stay a TAP-mode feature.
- **Detect whether the host has IPv6.** The default Docker network has *no* IPv6 (tested: no address, "Network unreachable"). The relay should test IPv6 connectivity at startup and **send no default route** in its advertisements if there isn't any. Otherwise guests wait on timeouts. When IPv6 is available, failed flows should get fast errors (TCP reset or ICMPv6 unreachable) so browsers fall back to IPv4 quickly.
- **ICMPv6 is required, not optional.** IPv6 routers never fragment, so "Packet Too Big" must work for PMTU discovery. The ICMP mechanisms above carry over directly: `IPPROTO_ICMPV6` ping sockets, and `IPV6_RECVERR`, whose "Packet Too Big" errors include the MTU. The relay must also compute ICMPv6 checksums.
- **Access rules: "block private ranges" doesn't protect the LAN in IPv6.** In IPv4 the operator's LAN is behind private ranges. In IPv6, LAN machines usually have **global** addresses from the ISP's prefix, so the rules need all of the following:
  - **Special-purpose ranges:** loopback, `fe80::/10` (link-local), `fc00::/7` (ULA; this includes e.g. Tailscale's `fd7a:115c:a1e0::/48`), `ff00::/8` (multicast), and the documentation and discard ranges.
  - **Ranges that hide an IPv4 address:** `::ffff:0:0/96` (IPv4-mapped), `64:ff9b::/96` (NAT64), `2002::/16` (6to4) and `2001::/32` (Teredo). Either decode the embedded IPv4 address and apply the IPv4 rules to it, or block the range. Otherwise e.g. `::ffff:192.168.1.1` would get around the IPv4 rules.
  - **The host's own global prefixes,** blocked automatically when the relay can see them.
  - **A new `WEBSOCKPROXY_BLOCKED_NETS` setting.** In a bridged container the relay can't see the host's prefixes, so operators must be able to list them. The docs need to say this clearly.
  - **IPv6 allowlist entries** in `WEBSOCKPROXY_ALLOWED_PRIVATE_NETS`.
- **Optional, later: NAT64/DNS64.** It's cheap in a socket-proxy design (`64:ff9b::a.b.c.d` becomes an IPv4 connection to `a.b.c.d`, and the DNS forwarder synthesises matching IPv6 answers), and would allow IPv6-only guests. Older guests and apps that use IPv4 literals would need extra client-side translation (464XLAT), so it should be opt-in.

### Running unprivileged: deployment changes
- **Container.** No `--privileged` and no `NET_ADMIN`.
  - **Tested:** in a default Docker container, even as `nobody` with `--cap-drop ALL`, the relay can bind port 80 (Docker sets `ip_unprivileged_port_start=0`) and open ping sockets.
  - The image can drop iptables, dnsmasq, iproute2 and bash, run as a non-root user, and use a read-only filesystem.
  - `docker-startup.sh` goes away in router mode.
- **Bare-host installs** need no root for router mode. At most, `net.ipv4.ping_group_range` must be widened for guest ping, and the relay needs permission to bind its port (or a non-privileged port).
- **Configuration:**
  - `WEBSOCKPROXY_UPLINK`.
  - Guest subnet and ULA prefix.
  - Upstream DNS server, overriding the host's resolver.
  - Per-client limits on flows, sockets and ICMP.
  - `WEBSOCKPROXY_ALLOWED_PRIVATE_NETS` and `WEBSOCKPROXY_BLOCKED_NETS`.
  - Whether to hide private traceroute hops.
- **Resource safety:** per-client and global flow caps, idle timeouts for each protocol, and bounded buffers (the downstream limits already added cover the websocket side).
- **Logging:** DHCP leases (MAC ↔ IP ↔ client) for abuse handling, and optional per-flow logging.

## How to build it: options for the network stack

1. **Pure Python.** No new dependencies, but the TCP implementation is weeks of work, and TCP edge cases are where bugs and security holes live. Only worth it if staying pure-Python is a hard requirement.
2. **Embed libslirp through a C binding (cffi). Recommended.**
   - **What it brings:** the stack QEMU and Podman use. It already provides ARP, DHCP, DNS forwarding, TCP/UDP NAT and ICMP echo (probably via ping sockets, falling back to raw sockets), and should support several guests on one virtual network, matching the switch model.
   - **The work:** the binding, plus connecting libslirp's timers and file-handle polling to asyncio.
   - **Check in a spike:** exactly how much of the ICMP error forwarding and IPv6 behaviour above libslirp already does. Any gaps can be filled alongside it, or fixed upstream.
3. **An external helper process such as passt,** speaking length-prefixed frames over a UNIX socket in place of the TAP device. It needs the least code in the relay, but adds a separately installed program. passt is also designed around a single guest (by default it gives the guest the host's own address), so it may not fit many clients sharing one network without one passt per client. Prototype before committing.

## Proposed phases

1. **Uplink abstraction and `WEBSOCKPROXY_UPLINK=none`**, the original scope of this issue, with tests.
2. **Spike:** a libslirp binding working with asyncio. Verify that several guests can share one network, and check its ICMP and IPv6 coverage against the lists above.
3. **IPv4 router uplink:** ARP, DHCP, DNS, UDP/TCP NAT, ICMP echo, access rules and per-client limits. Plus an unprivileged Docker image, and CI that runs `test_ping.py` (DHCP, ARP, DNS, ping) against it with `--cap-drop ALL`.
4. **ICMP error translation:** traceroute, PMTU and fast failures, with private-hop hiding.
5. **IPv6:** Neighbor Discovery, router advertisements with DNS, ULA addressing, detecting whether the host has IPv6, ICMPv6, the IPv6 access rules (including host-prefix blocking and `WEBSOCKPROXY_BLOCKED_NETS`) and IPv6 tests for `test_ping.py`.
6. **Optional:** NAT64/DNS64 and inbound port forwarding.

TAP mode stays supported throughout. If the router mode holds up in phases 3–5, it could become the default for the Docker image.
